### DISCUSSION TITLE
using every_second() outside of a driver in Berry

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#19198] [Q&A]
Is it possible to use these functions inside a berry program that's not a driver??

~~~
every_second(): called every second
every_50ms(): called every 50ms (i.e. 20 times per second)
every_100ms(): called every 100ms (i.e. 10 times per second)
every_200ms(): called every 200ms (i.e. 5 times per second)
every_250ms(): called every 250ms (i.e. 4 times per second)
~~~

**Comment by s-hadinger** [comment]
The only way to receive these calls is to register an instance with `tasmota.add_driver()`. That's all. 

It does not need to derive from any specific class. Berry uses duck typing, you just need a method with the adequate name

**Reply by sfromis** [reply]
You can also create a self-repeating timer at some arbitrary interval. [Example](https://tasmota.github.io/docs/Berry-Cookbook/#call-function-at-intervals) looks a bit complicated, but this is mostly due to it striving to avoid drift.
Or for whole seconds, you can use a cron timer.

**Reply by trlafleur** [reply]
is there a way to enlarge the berry console space to allow debugging a bigger program??

thanks

**Reply by trlafleur** [reply]
Allowing the use of the timing function calls in berry outside of a berry driver, would at first look to be more efficient in CPU resources than `cron or set_timer_modulo`

**Reply by s-hadinger** [reply]
I don't understand why you focus on "outside a berry driver". A driver is an abstract construct

**Reply by sfromis** [reply]
I'd suggest not worrying too much about marginal CPU efficiency until you have a measurement of the cost.

To me the "overhead" of the driver construct is not in CPU resources, but the need for removing and registering the driver for each test. This can easily be handled by coding like this:
```
class MyDriver
  def init()
    tasmota.remove_driver(global.mydriver_instance)
    global.mydriver_instance = self
    tasmota.add_driver(global.mydriver_instance)
  end
  def every_second()
    print('Tick Tock', tasmota.millis())
  end
end
```
And yes, this also works first time where the global var has not been created yet.

(Anyway, I'd recommend always doing the driver setup in the `init()` function instead of outside of the class).

**Reply by sfromis** [reply]
> is there a way to enlarge the berry console space to allow debugging a bigger program??

Getting a device with PSRAM does provide a lot more headroom for Berry.
